designs/sam_publish_app_cmd.rst

```diff
@@ -0,0 +1,339 @@
+.. contents:: **Table of Contents**
+   :depth: 2
+   :local:
+
+``sam publish app`` command
+====================================
+
+This is the design for a command to publish an application to `AWS Serverless Application Repository (SAR)`_ with a SAM
+template. It can be used to create a new application and its first version, update exisitng application's metadata, create
+a new version of the application, and manage application permissions.
+
+.. _AWS Serverless Application Repository (SAR): https://aws.amazon.com/serverless/serverlessrepo/
+
+
+What is the problem?
+--------------------
+To publish an app to AWS Serverless Application Repository, customers need to go through the following steps: first upload
+the application code and SAM template to an Amazon S3 bucket, correctly set S3 bucket policy that grants the service read
+permissions for artifacts uploaded to S3, then open the AWS Serverless Application Repository console and provide information
+in a bunch of input boxes. If they use the AWS CLI, they need to pass all the information as parameters, and it's easy to make
+a mistake while typing in the command line.
+
+
+What will be changed?
+---------------------
+In this proposal, we will be providing a new command, ``sam publish app``, which takes a SAM template as input and publishes
+an application to AWS Serverless Application Repository using applicaiton metadata specified in the template. Customers
+need to provide application metadata information in the template, then ``sam package`` will handle uploading local files to S3,
+and ``sam publish app`` will create the app in Serverless Application Repository. We will also provide the ``--acount-ids`` option
+to configure who the application can be shared with.
+
+
+Success criteria for the change
+-------------------------------
+#. Support all the following use cases:
+
+   * Create new application and its first version in SAR using ``sam publish app``
+   * Create new version of existing SAR application using ``sam publish app``
+   * Update application metadata of existing SAR application using ``sam publish app``
+   * Share the app publicly using ``--acount-ids ['*']``
```

Comment:
I strongly disagree with consolidating `--make-public`, `--make-private`, and `--account-ids` into just `--account-ids`. It degrades the customer experience.

Also, spelling: `--acount-ids` -> `--account-ids`

Reply:
We could add a confirmation on `--make-private`. I agree it's generally bad practice as a publisher, but it seems odd to make people use an entirely different mechanism, e.g., AWS CLI. If I accidentally made an app that was meant to be privately shared public, I'd want to undo it as quickly as possible.

Reply:
> Since the CLI is provided as a convenient tool, we probably shouldn't make it so easy to revoke shared apps.

Taking this a step further, then we shouldn't be changing permissions or operating on permissions.

> If I accidentally made an app that was meant to be privately shared public, I'd want to undo it as quickly as possible.

If you did this through another means, why wouldn't you do it through that means. If SAM CLI doesn't do permissions and we just publish whatever is in the template, this isn't an issue.

I see value in the use-case but do not see how it connects to publish. Publish parses the metadata and publishes the app or template to SAR. I don't think publish is the right flow or command to be managing permissions. Especially when they are not granular. These options provide connivence during publish but a version of an app doesn't have Permissions right? It is defined on the application itself. To me, we are doing to many things in publish and the scope grows from I need to get this into SAR to I need to get this into SAR and can manage things. I don't think publish is the right action for this use-case.

Reply:
An early design had `sam publish app` and `sam share app`, but permissions got merged into publish after an UX review. After discussing with James and Salman, we agreed that `sam publish app` is doing too many things, and we should take permissions out of scope. `sam share app` could be added as a follow-up.